### DISCUSSION TITLE
Dose preview mode only work with file-system based router?

### DISCUSSION THREAD

**Discussion by P233** [vercel/next.js#12126] [Help]
I am using the next.js custom server with koa in my project and trying to turn off `useFileSystemPublicRoutes`. But then I could not make the preview mode works properly.

In the custom server file I have this router for `api/preview`.
```js
app.prepare().then(() => {
  const server = new Koa();
  const router = new Router();

  router.get("/api/preview", async (ctx) => {
    await app.render(ctx.req, ctx.res, "/api/preview", ctx.query);
    ctx.respond = false;
  });  
  ...
});
```

This is the `/api/preview.ts` file: 
```js
export default async (req, res) => {
  res.setPreviewData({
    previewToken: req.query.token
  });
  ...
}
```

I got `TypeError: res.setPreviewData is not a function` and the `req` value is `{}`.

Now I have to use both custom server and enable `useFileSystemPublicRoutes`. Did I miss anything?

**Comment by timneutkens** [comment]
Would recommend not using a custom server and building around API routes to begin with
